### PR TITLE
e2e testing add tests for bulkpublishing

### PR DESCRIPTION
**Pull request by koepalex**
* Extended e2e test with bulk publishing test cases
  * orchestrated - will use the REST API of Industrial IoT Platform (tested with 2.7.199 and 2.7.206)
  * standalone - will replace published nodes json (tested with 2.8.142-alpha)
    * earlier released not working based on [dotnet/runtime issue 46978](https://github.com/dotnet/runtime/issues/46978)
* Extend existing TestHelper methods with list of expected OPC UA endpoints
* Added method to wait until expected endpoint is found to TestHelper
* Enforce ordering of test cases
  * each test theory has own priority range (50 step width) 
  * avoid async test methods

### CODE REVIEW

**Comment by mregen** [comment]
>                 "mcr.microsoft.com" : _context.ContainerRegistryConfig.ContainerRegistryServer;

is this a candidate for a string const?

---
Refers to: e2e-tests/IIoTPlatform-E2E-Tests/deploy/IoTHubPublisherDeployment.cs:64 in db0dbb0. [](commit_id = db0dbb00c5bc2fc8e05cf4c73c756220093b894c, deletion_comment = False)

File: e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/A_PublishSingleNodeOrchestratedTestTheory.cs [hunk 1]
```diff
@@ -14,56 +14,58 @@ namespace IIoTPlatform_E2E_Tests.Orchestrated
...
             Assert.NotEmpty(simulatedOpcServer.Values);
         }
 
         [Fact, PriorityOrder(3)]
-        public async Task Test_RegisterOPCServer_Expect_Success() {
+        public void Test_RegisterOPCServer_Expect_Success() {
 
             var cts = new CancellationTokenSource(TestConstants.MaxTestTimeoutMilliseconds);
 
             // We will wait for microservices of IIoT platform to be healthy and modules to be deployed.
```
Comment (mregen) [inline e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/A_PublishSingleNodeOrchestratedTestTheory.cs:63]: >We will wait  [](start = 15, length = 13)

in general its better to use a distant technical language than first person chat for comments

File: e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/PublishSingleNodeOrchestratedTestTheory.cs [hunk 2]
```diff
@@ -14,56 +14,58 @@ namespace IIoTPlatform_E2E_Tests.Orchestrated
...
             var cts = new CancellationTokenSource(TestConstants.MaxTestTimeoutMilliseconds);
-            var simulatedOpcServer = await TestHelper.GetSimulatedPublishedNodesConfigurationAsync(_context, cts.Token);
+            var simulatedOpcServer = TestHelper.GetSimulatedPublishedNodesConfigurationAsync(_context, cts.Token).GetAwaiter().GetResult();
             Assert.NotNull(simulatedOpcServer);
             Assert.NotEmpty(simulatedOpcServer.Keys);
             Assert.NotEmpty(simulatedOpcServer.Values);
         }
 
         [Fact, PriorityOrder(3)]
-        public async Task Test_RegisterOPCServer_Expect_Success() {
```
Comment (mregen) [inline e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/PublishSingleNodeOrchestratedTestTheory.cs:57]: >async Task [](start = 15, length = 10)

why not async tests?

File: e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/A_PublishSingleNodeOrchestratedTestTheory.cs [hunk 3]
```diff
@@ -303,26 +321,47 @@ public class PublishSingleNodeOrchestratedTestTheory
...
             var cts = new CancellationTokenSource(TestConstants.MaxTestTimeoutMilliseconds);
-            await Task.Delay(90 * 1000, cts.Token); //wait till the publishing has stopped
+            Task.Delay(90 * 1000, cts.Token).GetAwaiter().GetResult(); //wait till the publishing has stopped
             //use test event processor to verify data send to IoT Hub (expected* set to zero as data gap analysis is not part of this test case)
-            await TestHelper.StartMonitoringIncomingMessagesAsync(_context, 0, 0, 0, cts.Token);
+            TestHelper.StartMonitoringIncomingMessagesAsync(_context, 0, 0, 0, cts.Token).GetAwaiter().GetResult();
             // wait some time to generate events to process
-            await Task.Delay(90 * 1000, cts.Token);
-            var json = await TestHelper.StopMonitoringIncomingMessagesAsync(_context, cts.Token);
+            Task.Delay(90 * 1000, cts.Token).GetAwaiter().GetResult();
```
Comment (mregen) [inline e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/A_PublishSingleNodeOrchestratedTestTheory.cs:341]: >90 * 1000 [](start = 23, length = 9)

candidate for a 'common test parameter, I see similar timeouts in various places.

File: e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/A_PublishSingleNodeOrchestratedTestTheory.cs [hunk 4]
```diff
@@ -303,26 +321,47 @@ public class PublishSingleNodeOrchestratedTestTheory
...
+                Assert.True(response.IsSuccessful, "DELETE /publisher/v2/jobs/{jobId} failed!");
             }
         }
 
         [Fact, PriorityOrder(12)]
-        public async Task Test_VerifyNoDataIncomingAtIoTHub() {
+        public void Test_VerifyNoDataIncomingAtIoTHub() {
             var cts = new CancellationTokenSource(TestConstants.MaxTestTimeoutMilliseconds);
-            await Task.Delay(90 * 1000, cts.Token); //wait till the publishing has stopped
+            Task.Delay(90 * 1000, cts.Token).GetAwaiter().GetResult(); //wait till the publishing has stopped
```
Comment (mregen) [inline e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/A_PublishSingleNodeOrchestratedTestTheory.cs:337]: >.GetAwaiter().GetResult() [](start = 44, length = 25)

this is really ugly, there must be a good explanation why async can not be used...

File: e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/B_PublishMultipleNodesOrchestratedTestTheory.cs [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+
+        [Fact, PriorityOrder(54)]
+        public void Test_VerifyDataAvailableAtIoTHub() {
+
+            var cts = new CancellationTokenSource(TestConstants.MaxTestTimeoutMilliseconds);
+
+            //use test event processor to verify data send to IoT Hub (expected* set to zero as data gap analysis is not part of this test case)
+            TestHelper.StartMonitoringIncomingMessagesAsync(_context, 50, 1000, 90_000_000, cts.Token).GetAwaiter().GetResult();
+            // wait some time to generate events to process
+            Task.Delay(180 * 1000, cts.Token).GetAwaiter().GetResult();
```
Comment (mregen) [inline e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/B_PublishMultipleNodesOrchestratedTestTheory.cs:260]: >180 * 1000 [](start = 23, length = 10)

magic number

File: e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/B_PublishMultipleNodesOrchestratedTestTheory.cs [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+                _output.WriteLine($"StatusCode: {response.StatusCode}");
+                _output.WriteLine($"ErrorMessage: {response.ErrorMessage}");
+                Assert.True(response.IsSuccessful, "POST /publisher/v2/publish/{endpointId}/bulk failed!");
+            }
+        }
+
+        [Fact, PriorityOrder(56)]
+        public void Test_VerifyNoDataIncomingAtIoTHub() {
+            var cts = new CancellationTokenSource(TestConstants.MaxTestTimeoutMilliseconds);
+            Task.Delay(90 * 1000, cts.Token).GetAwaiter().GetResult(); //wait till the publishing has stopped
```
Comment (mregen) [inline e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/B_PublishMultipleNodesOrchestratedTestTheory.cs:326]: >90 * 1000 [](start = 23, length = 9)

magic number

File: e2e-tests/IIoTPlatform-E2E-Tests/Standalone/A_PublishSingleNodeStandaloneTestTheory.cs [hunk 7]
```diff
@@ -16,92 +16,89 @@ namespace IIoTPlatform_E2E_Tests.Standalone {
...
             var cts = new CancellationTokenSource(TestConstants.MaxTestTimeoutMilliseconds);
 
             //use test event processor to verify data send to IoT Hub (expected* set to zero as data gap analysis is not part of this test case)
-            await TestHelper.StartMonitoringIncomingMessagesAsync(_context, 0, 0, 0, cts.Token);
+            TestHelper.StartMonitoringIncomingMessagesAsync(_context, 0, 0, 0, cts.Token).GetAwaiter().GetResult();
 
             // wait some time to generate events to process
-            await Task.Delay(90 * 1000, cts.Token);
-            var json = await TestHelper.StopMonitoringIncomingMessagesAsync(_context, cts.Token);
+            Task.Delay(90 * 1000, cts.Token).GetAwaiter().GetResult();
```
Comment (mregen) [inline e2e-tests/IIoTPlatform-E2E-Tests/Standalone/A_PublishSingleNodeStandaloneTestTheory.cs:79]: >90 * 1000 [](start = 23, length = 9)

magic number

**Review by mregen — APPROVED** [review]
:shipit:

**Comment by mregen** [comment]
now with LGTM :-)

File: e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/PublishSingleNodeOrchestratedTestTheory.cs (hunk 2 above)
Comment (koepalex) [reply e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/PublishSingleNodeOrchestratedTestTheory.cs:57]: Test are modifying a critical resource (Test Environment) why running (async) the behavior is non deterministic and we can't understand who the system is changed. Under this conditions it is not possible to run reliable tests.

File: e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/A_PublishSingleNodeOrchestratedTestTheory.cs (hunk 3 above)
Comment (koepalex) [reply e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/A_PublishSingleNodeOrchestratedTestTheory.cs:341]: used existing constant TestConstants.DefaultTimeoutInMilliseconds

File: e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/A_PublishSingleNodeOrchestratedTestTheory.cs (hunk 4 above)
Comment (koepalex) [reply e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/A_PublishSingleNodeOrchestratedTestTheory.cs:337]: see https://github.com/Azure/Industrial-IoT/pull/1020#discussion_r563553027

File: e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/B_PublishMultipleNodesOrchestratedTestTheory.cs (hunk 5 above)
Comment (koepalex) [reply e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/B_PublishMultipleNodesOrchestratedTestTheory.cs:260]: true; added comment why the magic number is used

File: e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/B_PublishMultipleNodesOrchestratedTestTheory.cs (hunk 6 above)
Comment (koepalex) [reply e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/B_PublishMultipleNodesOrchestratedTestTheory.cs:326]: no, this is the default timeout, changed to use constant TestConstants.DefaultTimeoutInMilliseconds

File: e2e-tests/IIoTPlatform-E2E-Tests/Standalone/A_PublishSingleNodeStandaloneTestTheory.cs (hunk 7 above)
Comment (koepalex) [reply e2e-tests/IIoTPlatform-E2E-Tests/Standalone/A_PublishSingleNodeStandaloneTestTheory.cs:79]: no, this is the default timeout, changed to use constant TestConstants.DefaultTimeoutInMilliseconds

File: e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/A_PublishSingleNodeOrchestratedTestTheory.cs [hunk 8]
```diff
@@ -14,56 +14,58 @@ namespace IIoTPlatform_E2E_Tests.Orchestrated
...
+            _context.Reset();
+            TestHelper.SwitchToOrchestratedModeAsync(_context).GetAwaiter().GetResult();
         }
 
         [Fact, PriorityOrder(1)]
-        public async Task Test_CollectOAuthToken() {
+        public void Test_CollectOAuthToken() {
             var cts = new CancellationTokenSource(TestConstants.MaxTestTimeoutMilliseconds);
-            var token = await TestHelper.GetTokenAsync(_context, cts.Token);
+            var token = TestHelper.GetTokenAsync(_context, cts.Token).GetAwaiter().GetResult();
```
Comment (barnstee) [inline e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/A_PublishSingleNodeOrchestratedTestTheory.cs:45]: Definitely needs a comment why this is not async: "XUnit executes predictively when tests are running with synchronous/blocking threads."

File: e2e-tests/IIoTPlatform-E2E-Tests/TestHelper.cs [hunk 9]
```diff
@@ -471,16 +480,119 @@ private static PrivateKeyFile GetPrivateSshKey(IIoTPlatformTestContext context)
                     }
 
                     Assert.NotEmpty(response.Content);
-                    json = JsonConvert.DeserializeObject(response.Content);
+                    json = JsonConvert.DeserializeObject<ExpandoObject>(response.Content, new ExpandoObjectConverter());
```
Comment (barnstee) [inline e2e-tests/IIoTPlatform-E2E-Tests/TestHelper.cs:483]: ExpandedObject?

**Review by barnstee — APPROVED** [review]

File: e2e-tests/IIoTPlatform-E2E-Tests/TestHelper.cs (hunk 9 above)
Comment (koepalex) [reply e2e-tests/IIoTPlatform-E2E-Tests/TestHelper.cs:483]: Avoid an error in Newtonsoft.Json deserialization, ExpandoObject are dynamic classes  were the calls to the object are verified at runtime. [MSDON ExpandoObject](https://docs.microsoft.com/en-us/dotnet/api/system.dynamic.expandoobject?view=netcore-3.1)

File: e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/A_PublishSingleNodeOrchestratedTestTheory.cs (hunk 8 above)
Comment (koepalex) [reply e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/A_PublishSingleNodeOrchestratedTestTheory.cs:45]: see https://github.com/Azure/Industrial-IoT/pull/1020#discussion_r563553027

File: e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/A_PublishSingleNodeOrchestratedTestTheory.cs (hunk 1 above)
Comment (koepalex) [reply e2e-tests/IIoTPlatform-E2E-Tests/Orchestrated/A_PublishSingleNodeOrchestratedTestTheory.cs:63]: comment is not new, will be changed in further version

**Pull request merged**